OHM-871 BAHMNI Mapping Functions

MattyJ007 opened the pull request:
Dubnium is the LTS and testing latest node is useful.

The previous version 0.10 being tested is no longer relevant

OHM-871

MattyJ007 committed:
Increase the node versions tested

Dubnium is the LTS and testing latest node is useful.

The previous version 0.10 being tested is no longer relevant

OHM-871

MattyJ007 committed:
Add function createBahmniFhirIdentifier

This function will create a FHIR identifier from a Bahmni patient record

OHM-871

MattyJ007 committed:
Add function to create FHIR name for Bahmni

Transform the Bahmni patient name fields into the FHIR patient name
fields.

OHM-871

MattyJ007 committed:
Add function to map FHIR gender from a code

Bahmni patient gender is determined by a single character code.

FHIR expects one of the following words:
male, female, other, unknown

OHM-871

MattyJ007 committed:
Add function to translate address format

This funciton will translate a Bahmni address into a FHIR 3 address
format

OHM-871

Zooloo2014 committed:
Add happy case test for transform dateTimeToDate OHM-871

MattyJ007 committed:
Remove unused custom functions

Those functions were needlessly complex as the object-mapper core engine
could accomplish those mappings with the correct syntax in the mapping
schema.

OHM-871

MattyJ007 committed:
Merge branch 'OHM-871-bahmni-mapping-functions' of github.com:jembi/node-object-mapper into OHM-871-bahmni-mapping-functions

MattyJ007 committed:
Add generic function to flip boolean values

Useful in cases where the mapped fields correspond with inverse boolean
values.
Eg: if the input field name is "active" and the output field is
"inactive". The field value just requires the negative boolean to map
correctly

OHM-871

MattyJ007 committed:
Split out tranformation fuction into separate files

To make maintaining the functions easier.
No monoliths please...

OHM-871

MattyJ007 committed:
Remove duplicate transformation test

Datetime to date test is alredy there

OHM-871

MattyJ007 committed:
Fix test syntax

Firstly, the tests ran using "node" not "tape" and they only ran
the test.js file which excluded the tests for transformations.

Secondly, the reference to the tests to be tested weren't correct.

OHM-871

MattyJ007 committed:
Add tests for boolean flip function

Check function acts as expected

OHM-871

MattyJ007 committed:
Refactor project to reflect Jembi's ownership

The project has been modified from its fork origin therefore these
chagnes need to be represented.

All references to previous owner in the code in liceses and contacts
will be removed.

New code features and how to guides will be updated

Reference new repo location

OHM-871

MattyJ007 committed:
Update installation instruction

We don't have an NPM release of the project therefore to get the code it
can be referenced from our github repo

OHM-871

MattyJ007 committed:
Fix Readme title

Name was not accurate

OHM-871

MattyJ007 committed:
Add tests for gender code map function

Test it works as expected

OHM-871

BMartinos commented on package.json:
```diff
@@ -1,39 +1,23 @@
-  "description": "Copy properties from one object to another.",
-  "version": "5.0.0",
+  "description": "Map fields from one object to another user a user defined schema",
+  "version": "1.0.0",
```
Lets maybe start this at `0.1.0` as an initial version and once we are happy with everything we can bump to an official 1 release.

BMartinos commented on README.md:
````diff
@@ -9,8 +9,10 @@ on a separate `Object`, which defines how the properties should be mapped.
+In your projects `package.json` file include the following line in the dependencies field:
+
+```json
+"object-mapper": "jembi/node-object-mapper#master"
````
Do we need to supply the hash for master? would it not default to master in any case?

BMartinos commented on README.md:
```diff
@@ -68,30 +71,44 @@ var map = {
-      transform: function (value) { 
-        return value + "_foo";
+      transform:
+        {
```
Maybe just for consistency lets move the brackets after the colon instead of a new line

BMartinos commented on src/transform/codes.js:
```diff
@@ -0,0 +1,19 @@
+      default:
+        return 'unknown'
+    }
+  }
```
As mentioned to you in a chat, it would be great if we could get rid of these hard coded cases which might become problematic when adding more cases for other input values. The idea we could work on is to have these cases built up based on the user supplied parameters set in the `input-mapping.json`. This will make it more flexible to the implementing user and less maintenance on the developers.

BMartinos commented on src/transform/primitives.js:
```diff
@@ -0,0 +1,14 @@
+    }
+
+    return null
+  }
```
Maybe as a future add, but we add add another function here to do a `booleanConvert` function. This will be useful in the case where a value is a string but meant to be Boolean. This will take in the value supplied and make a boolean out of it. The below example should return a boolean of `false`
E.g:
```
{
  booleanValue: "false"
}
```

BMartinos replied:
This will become important when we can start chaining function together. So if the value is a string of false, but we need to populate it as a Boolean true we can chain the `booleanConvert` first and then the `booleanFlip` functions

BMartinos commented on src/transform/tests.js:
```diff
@@ -0,0 +1,45 @@
+    dst.manual = val.a + val.b
+  },
+  'test-empty-array': function(sourceValue) {
+    var i
```
Not sure if this was exiting or not, but I think we can remove this declaration here and do it in the for loop instead.

MattyJ007 replied:
Lol, that was from the original tests - went straight from eyes to fingers by-passing brain

BMartinos reviewed (changes requested):
Thanks @MattyJ007 looking good so far, just a few minor comments

MattyJ007 committed:
Reduce Application version

The version reflected here shoudl indicate that this is not a full
release yet - we're still developing our features

OHM-871

MattyJ007 committed:
Remove git branch specification in install

When installing a dependency from github the install will use master by
default therefore it isn't necessary to specify the master branch

OHM-871

MattyJ007 committed:
Fix bracket spacing

Boyscout

OHM-871

MattyJ007 committed:
Add function to convert user values to boleans

Generally a useful mapping functionality

OHM-871

MattyJ007 committed:
Imporve test for loop declaration

Boyscout

OHM-871

MattyJ007 committed:
Add functionParameters parameter to transform function

This parameter will make a set of user defined parameters available to
the transform function.

The purpose is to allow the user to specify a set of data to be used in
our preconfigured functions. For example, the mapping of codes to values
will require the user to provide a mapping of codes to values which we
will then use to map fields inplementing this function.

This makes the openhim-mediator-mapping service more configurable by the
user.

OHM-878

MattyJ007 committed:
Add code mapping function accepting user parameters

This versatile function will allow users to provide the map of codes to
values taht we will use for specific fields that require code to value
mappings

The use case in the openhim-mediator-mapping currently if converting
gender character codes to words to be accepted by the target system
(FHIR)

OHM-871

MattyJ007 committed:
Fix indentation

Consistency...

OHM-871

MattyJ007 committed:
Add tests for Booleanify function

Boyscout :p

OHM-871

BMartinos commented on README.md:
```diff
@@ -210,60 +209,60 @@ mapping defined by **mapObject**
+- **sourceObject** is the object FROM which properties will be copied.
+- **destinationObject** [OPTIONAL] is the object TO which properties will be copied.
+- **mapObject** is the object which defines how properties are copied from
+  **sourceObject** to **destinationObject**
```
Should this not be on the same line?

BMartinos reviewed (approved):
LGTM :+1:

BMartinos commented on test/transformationsTest.js:
```diff
@@ -0,0 +1,291 @@
+      t.end()
+    })
+
+    t.test('should convert unhandled code to null', t => {
```
Im not sure this test title reflects the actual test case. It seems to be testing the logic same as the previous test

MattyJ007 replied:
:| Whoops

BMartinos commented on test/transformationsTest.js:
```diff
@@ -0,0 +1,291 @@
+      'should convert unhandled codes to null if no default specified',
+      t => {
+        const inputCode = 7
+        const expectedOutput = undefined
```
Should this be `undefined` or `null`?

MattyJ007 replied:
Undefined :| 
I'm not sure if that will be an issue or not for object-mapper... I would assume not?

BMartinos replied:
Im not to sure but dont think so, just need to update the test title to state `undefined` insted of `null` :)

BMartinos commented on test/transformationsTest.js:
```diff
@@ -0,0 +1,291 @@
+    )
+
+    t.test('should return null if input is a falsy', t => {
+      const booleanInput = 0
+      const expectedOutput = null
```
Would this not be returning `false` instead of `null`?
https://github.com/jembi/node-object-mapper/pull/1/files#diff-0dec99e818a2242edae67563675d4d06R25

BMartinos replied:
oh wait nvm, thats a different function call...

MattyJ007 replied:
Nope - this is testing the booleanFlip function which won't accept truthys and falsys.

booleanify will change 0 to false

MattyJ007 committed:
Fix readme line spacing

Boyscout

MattyJ007 committed:
Update function name in test description

The function being tested had a name change :(

OHM-871

MattyJ007 committed:
Remove duplicate test

This test basically does that same check as the test that follows - but
it's name isn't as descriptive

OHM-871

MattyJ007 committed:
Correct test description

The description did not reflect the content of the test accurately

OHM-871

MattyJ007 committed:
Fix typo in package.json project description

Boyscout

OHM-871
